sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,605 @@
+attribute:
+
+* Declares a variable of that type.
+* Has a formal parameter declared with that type.
+* Returns that type.
+
+This also includes any qualified version of the type.
+
+**TODO**: This language is not very precise.  The intent is to include most
+uses of the type, except for cases like `sizeof(T)` or `decltype(T)`.  Help
+appreciated on tightening the wording here.
+
+**TODO**: Would it be better to use a different attribute name when decorating
+types, rather than expanding the meaning of `[[sycl::requires()]]`?  If we did
+this, the new attribute would become an internal DPC++ implementation detail;
+we would not add it to the SYCL specification.
+
+### Implementing diagnostics in the DPC++ front-end
+
+As noted above, the front-end device compiler must behave as though there is an
+implicit `[[sycl::requires(has(aspect::fp64))]]` attribute on any use of the
+`double` type in device code.
+
+Aside from this, the front-end compiler can implement the required diagnostics
+purely from the C++ attributes that have been added to the DPC++ headers.
+There is no need for the front-end compiler to know which device features are
+optional.
+
+When the front-end compiler sees a kernel or device function that is decorated
+with `[[sycl::requires()]]`, it forms the set of allowed aspects for that
+kernel or device function using aspects listed in the attribute.  Let's call
+this the `Allowed` set.  The front-end then computes the static call tree of
+that kernel or device function and forms the union of all aspects in any
+`[[sycl::requires()]]` attributes that decorate any of these functions or any
+of the types used inside these functions.  Let's call this the `Used` set.  If
+the `Used` set contains any aspects not in the `Allowed` set, the front-end
+issues a diagnostic.
+
+In order to be user-friendly, the diagnostic should point the user to the
+location of the problem.  Therefore, the diagnostic message should include the
+following information:
+
+* The source position of the `[[sycl::requires()]]` attribute that decorates
+  the kernel or device function which provides the `Allowed` aspect set.  This
+  tells the user which aspects the kernel or device function intends to use.
+
+* The source position of the call to a function that is decorated with
+  `[[sycl::requires()]]` or the source position of the use of a type that is
+  decorated with `[[sycl::requires()]]`.  This tells the user where in the
+  kernel a particular aspect is actually used.
+
+Note that this analysis can be done in the front-end compiler even when a
+kernel makes a call to a function that is in another translation unit.
+Language rules require the application to declare such a function with
+`SYCL_EXTERNAL` in the calling TU, and the `SYCL_EXTERNAL` declaration must be
+decorated with the `[[sycl::requires()]]` attribute.  Therefore, the front-end
+can diagnose errors with aspect usage even without seeing the definition of the
+`SYCL_EXTERNAL` function.
```

Comment:
This is only a partial solution, IMO. Nothing prevents a user from just writing whatever attributes shut the compiler up, regardless of what's in the other TU. While that's undoubtedly UB, it's also UB that's likely to "work" just enough to catch users out. Also, a bigger issue is the ergonomics of requiring cross-TU collusion like that -- it's easy to update one place in the code and forget the others.

IMO, a better approach to this is to use a static analysis check with cross-TU support.

Reply:
I think there are two concerns here:

1. You are concerned that the compiler cannot adequately diagnose problems with `[[sycl::requires()]]` when attached to `SYCL_EXTERNAL` functions without doing inter-TU analysis.

2. You are concerned that the FE should not be required to construct a static call tree in order to diagnose an error, and you think this sort of analysis should be exposed via a separate analyzer tool (which could also do inter-TU analysis).

I'll start by addressing (1), but I think (2) is your larger concern.

My expectation is that application that decorate `SYCL_EXTERNAL` functions with `[[sycl::requires()]]` will place this decoration in a header file that is included by both the TU that calls the `SYCL_EXTERNAL` function and by the TU that defines the `SYCL_EXTERNAL` function.  For example, the header might look like:

```
[[sycl::requires(has(aspect::fp16))]]
SYCL_EXTERNAL void do_16bit_thing();
```

When the header is included by the TU that defines `do_16bit_thing()`, the compiler will check that the function does not use any other features other than `aspect::fp16`.  When the header is included by the TU that calls `do_16bit_thing()`, the compiler assumes that the function uses `aspect::fp16` without doing any checks.  Therefore, usage like this is safe even though the compiler does not do any inter-TU analysis.

Of course, the developer can shoot themselves in the foot by declaring the function directly in the "*.cpp" file without using a shared header.  In such a case, there is no guarantee that the `SYCL_EXTERNAL` declaration matches the definition.  I don't think we need to be overly concerned, though, about code like this that goes out of its way to shoot itself in the foot.

I think your larger concern is (2), that the clang community will not be happy about adding diagnostics in the FE that require construction of a static call tree.  Obviously, we don't want to make changes which cannot be upstreamed.  My main concern about moving this diagnostic to a separate analysis tool is that customers may not run that tool often.  As a result, I fear that incorrect programs will often go undiagnosed.

It seems to me, though, that we must currently build a static call tree for each kernel in the device compiler.  I believe the device compiler identifies all the device code by building the static call tree of each kernel (and each `SYCL_EXTERNAL` function) and then embedding the LLVM IR for the device code in the fat binary.

Therefore, it may be more appropriate to add this diagnostic message to the device compiler pass.  If that pass builds the static call tree anyways, it should be straightforward to add support for this diagnostic.

I need to understand how the device compiler builds the static call tree today.  Is that done in the compiler front-end, so I should ask you @AaronBallman, or is it done later in the compiler and I should ask someone else?

Reply:
> My main concern about moving this diagnostic to a separate analysis tool is that customers may not run that tool often. As a result, I fear that incorrect programs will often go undiagnosed.

I just want to make sure we're all on the same page because I've heard this concern raised before and it raises red flags for me. There's no separate tool involved, this is all "just clang". Clang has lexing/parsing/semantic/codegen diagnostics that are emitted by the frontend -- those are your typical diagnostics, like: https://godbolt.org/z/rzYocYfqs. It also has diagnostics that are emitted by analysis passes -- those are ones that require control flow analysis but typically don't require interprocedural analysis. You have to manually enable these diagnostics because they slow down compilation somewhat, but you enable them with warning flags, like: https://godbolt.org/z/j864rzzo5. It also has diagnostics that are emitted by static analysis passes -- those are ones that require more complicated data and flow control analysis, including interprocedural analysis. You have to manually enable these as well, and there are separate tools (scan-build, CodeChecker, etc) that makes it easier to enable these as part of the build process, but it's not required, like: https://godbolt.org/z/5deM15YsT. But in all of the cases, it's "just clang" doing the lifting. There's also clang-tidy, which is run either through the static analyzer or through a separate executable, and it does AST pattern matching-based checking. My suggestion is that we should be implementing more of this logic as [static] analysis passes because that's more where this sort of functionality is expected to live in the toolchain, though I don't think we should discount clang-tidy (*especially* for lint-like functionality).

My viewpoint is that the SYCL language spec should not require anything other than statically known information (roughly: information that's carried by the AST node(s) involved in the construct without any flow analysis) to generate diagnostics for code that renders the program ill-formed (e.g, errors). This is the strategy taken by the C and C++ language designers as well and it makes for a more reasonable implementation burden for language features. For other situations where we really wish we could make it an error but can't because it requires further analysis (control or data flow or cross translation unit), the standard makes those cases "ill-formed, no diagnostic required" so that implementations don't have to go to heroics to diagnose. Alternatively, they make them undefined behavior so that implementations can do whatever they want (define the behavior, diagnose the code if they want to, don't diagnose the code if they don't want to, etc). While it'd be great to diagnose all of these IFNDR or UB situations, the expectation should be that implementations *won't* diagnose in those situations until there's a clear need to go to those efforts (and even then, the implementation may not go to those lengths). I think some of the current error situations should likely be IFNDR rather than requiring a diagnostic, and then we can make them into warnings (that are potentially upgraded to an error by default) in the appropriate analysis pass.

For warnings, I don't know that the fear is anything we can do something about because users control those. Nothing prevents a user from disabling all warning output with `-Wno-everything`, as an example. But my expectation would be that we'd want something like `-Wsycl` as a warning flag that would enable all of the SYCL-related warnings, including the analysis ones (but not the static analysis ones). If we have things that require cross-TU analysis, rely on taint analysis, etc, we may still have to push them down into the static analyzer and live with the fact that not everyone will run that tool.

> It seems to me, though, that we must currently build a static call tree for each kernel in the device compiler. I believe the device compiler identifies all the device code by building the static call tree of each kernel (and each SYCL_EXTERNAL function) and then embedding the LLVM IR for the device code in the fat binary.

Yes, that currently happens in the FE and is what I'm pushing back of adding more diagnostic logic to. IMO, it's not sustainable for diagnostics even if it’s needed for codegen purposes. The fact that some of this is used for *errors* is especially concerning because it makes error logic mysterious (if the call graph gets constructed one way, you get the error, but if you perturb the code sufficiently, the error goes dormant while still remaining in the user's source code).

Reply:
OK, how about this as a high level design strategy:

* The FE will just be responsible for decorating the LLVM IR with [IR function attributes](https://llvm.org/docs/LangRef.html#function-attributes) which capture the information about the `[[sycl::requires()]]` C++ attributes and information about the "used" aspects.  (I know that I have not defined "used" well yet.  I still plan to respond to that comment.)  There is another comment thread in this PR where we discuss using IR function attributes to convey information to the post-link tool.

* The FE will not be required to raise diagnostics when a device function uses an optional feature that is not listed in a `[[sycl::requires()]]` attribute.  Instead, these diagnostics will be added to either the clang analysis pass or the clang static analysis pass as you suggest above.  Since these passes aren't enabled by default, users won't get these diagnostic checks unless they specifically ask for them.

* The post-link tool also needs to build each kernel's static call tree, so we can add diagnostic checks there.  If an error is found, the diagnostic message will probably not contain the source position of the offending code.  Instead, it will just contain the name of the offending function.  However, the diagnostic message can tell the user how to enable the checking in the analysis pass, which will provide a better error message.

I think this achieves your goal of not wanting to perform heroic diagnostic checking in the FE, while also achieving my goal of wanting to inform users when their code uses optional features incorrectly.

Reply:
I think this sounds like a good path forward, but I want to make sure I'm clear on one point: 

> Instead, these diagnostics will be added to either the clang analysis pass or the clang static analysis pass as you suggest above.

Are these diagnostics warnings or errors? If they're warnings, then I think this'll work well. If they're errors, then we're drifting somewhat into novel territory of having a flag to enable errors. If you really need them to be errors, one possible compromise approach would be to treat them as warnings that default to an error. They can still be manually disabled to ignore the error, but they're pretty error-like in the common case. e.g., https://godbolt.org/z/xM6zrsE17

Reply:
The SYCL spec doesn't distinguish between errors and warnings; it just says that a "diagnostic" is required.  I think a good analogy would be the C++ `[[noreturn]]` attribute.  What does clang do if a function is decorated with this attribute, and the compiler can detect that the function does return?  Is this an error or a warning?

I think the behavior of `[[sycl::requires()]]` should be similar.  Like `[[noreturn]]`, this is also an assertion by the programmer that a function has a certain characteristic.

Reply:
> The SYCL spec doesn't distinguish between errors and warnings; it just says that a "diagnostic" is required. 

This is a bit of an aside that's beyond your proposal here, but that seems like something you may want to address in the SYCL spec. In C++, programs are either ill-formed or well-formed. If the program is well-formed, then an implementation must translate the program. The same is not true if the program is ill-formed. The basic idea is that when a program is ill-formed, the user will get an error message and otherwise the standard doesn't *require* any diagnostics whatsoever, but it might *recommended* a warning in some circumstances.

I'd imagine SYCL wants a similar distinction, otherwise a conforming (low-quality) implementation can write "Congratulations, you've used SYCL" as a "diagnostic message" and that covers all semantic checking requirements for the SYCL spec. Having some delineation makes it clear to implementers what the expectations are.

> I think a good analogy would be the C++ [[noreturn]] attribute. What does clang do if a function is decorated with this attribute, and the compiler can detect that the function does return? Is this an error or a warning?

That's recommended to be a warning (http://eel.is/c++draft/dcl.attr.noreturn#3).

> I think the behavior of [[sycl::requires()]] should be similar. Like [[noreturn]], this is also an assertion by the programmer that a function has a certain characteristic.

Okay, I think that's a good answer here because it means we can implement this as a warning in whatever part of the compiler makes the most sense, which could include not warning the user at all in circumstances that are too hard to identify.